source/common/router/router.cc

```diff
@@ -580,7 +580,11 @@ Http::FilterHeadersStatus Filter::decodeHeaders(Http::RequestHeaderMap& headers,
       (upstream_http_protocol_options.value().auto_sni() ||
        upstream_http_protocol_options.value().auto_san_validation())) {
     // Default the header to Host/Authority header.
-    absl::string_view header_value = headers.getHostValue();
+    std::string header_value = route_entry_->getRequestHostValue(headers);
+    if (!Runtime::runtimeFeatureEnabled(
```

Comment:
so this seems plausible and safely done, but out of curiousity, is it possible to instead of factoring out the host header rewrite logic and calling it twice, to factor out this entire if statement into a function and (runtime guard) move it below where we finalizeReuqestHeaders and just snag the host header there?

Reply:
I guess it should be possible, my initial intent was to keep the original sequence of operations as much as possible to be safe.
Tomorrow I will move the SNI/SAN portion after the header is finalized and see if there is any regression.

Reply:
I tried but it did not work as hope (https://github.com/envoyproxy/envoy/pull/35566/commits/ba4af6253544c6f1931d06788304d317cee4d778)
Moving the configuration of SNI and SAN after the header are finalized will not take effect as that configuration needs to be in place before the createConnPoll is called. https://github.com/envoyproxy/envoy/blob/ba4af6253544c6f1931d06788304d317cee4d778/source/common/router/router.cc#L597

The only remaining option to try is to move finalizeRequestHeaders earlier in the flow. However, I don´t think that is a viable option due to the "destructive" changes it performs on the headers.

I will revert the last commit for now